rllib/algorithms/ppo/utils/ppo_env_runner.py

```diff
@@ -0,0 +1,561 @@
+                force_reset=False,
+            )
+        # Sample m episodes.
+        else:
+            # `_sample_episodes` returns only a single list (with completed episodes)
+            # therefore add the empty list for the truncated episodes.
+            return self._sample_episodes(
+                num_episodes=num_episodes,
+                explore=explore,
+                random_actions=random_actions,
+                with_render_data=with_render_data,
+            )
+
+    def _sample_timesteps(
+        self,
+        num_timesteps: int,
+        explore: bool = True,
+        random_actions: bool = False,
+        force_reset: bool = False,
+    ) -> List[Episode]:
+        """Helper method to sample n timesteps."""
+
+        done_episodes_to_return = []
+
+        # Get initial states for all 'batch_size_B` rows in the forward batch,
+        # i.e. for all vector sub_envs.
+        if hasattr(self.marl_module, "get_initial_state"):
+            initial_states = tree.map_structure(
+                lambda s: np.repeat(s, self.num_envs, axis=0),
+                self.marl_module[DEFAULT_POLICY_ID].get_initial_state(),
+            )
+        else:
+            initial_states = {}
+
+        # Have to reset the env (on all vector sub_envs).
+        if force_reset or self._needs_initial_reset:
+            obs, infos = self.env.reset()
+
+            self._episodes = [Episode() for _ in range(self.num_envs)]
+            states = initial_states
+
+            # Set initial obs and states in the episodes.
+            for i in range(self.num_envs):
+                # Extract info for the vector sub_env.
+                self._episodes[i].add_initial_observation(
+                    initial_observation=obs[i],
+                    initial_info=infos[i],
+                    # TODO (simon): Check, if this works for the default
+                    # stateful encoders.
+                    initial_state={k: s[i] for k, s in states.items()},
+                )
+        # Do not reset envs, but instead continue in already started episodes.
+        else:
+            # Pick up stored observations and states from previous timesteps.
+            obs = np.stack([eps.observations[-1] for eps in self._episodes])
+            # Compile the initial state for each batch row (vector sub_env):
+            # If episode just started, use the model's initial state, in the
+            # other case use the state stored last in the Episode.
+            states = {
+                k: np.stack(
+                    [
+                        initial_states[k][i] if eps.states is None else eps.states[k]
+                        for i, eps in enumerate(self._episodes)
+                    ]
+                )
+                for k in initial_states.keys()
+            }
+
+        # Loop through env in enumerate.(self._episodes):
+        ts = 0
+        pbar = tqdm(total=num_timesteps, desc=f"Sampling {num_timesteps} timesteps ...")
```

Comment:
Dumb question: I think tqdm is currently not a requirement for ray/rllib.
Also, would this work on n remote workers as well?